Ensure all icons in bookmarks bar are vertically aligned

petemill opened the pull request:
Ensure all icons in bookmarks bar are vertically aligned
Fix #13789

Also restore different bookmarks bar height when in text-only mode vs with favicons.
Fix #13985

Reduce flicker of bookmark bar / overflow-indicator / last item by having aphrodite not remove the custom class-name on render.

Reduce complexity of bookmarks toolbar state by passing through ToolbarMode 'enum' to component.
Future-proof bookmark mode setting retrieval for state-selector-memoization by passing through appState

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

## Test Plan:
1. Preferences > Bookmarks Bar > choose `Show favicons only`
2. Add 10 - 20 bookmarks to bookmarks bar (or pre-existing profile)
3. Bookmark the New Tab page (as it does not have a favicon)
4. Add a couple of folders to bookmarks bar
5. Extend window width so that there is no overflow indicator
6. Contract window width so that there is an overflow indicator
   - **Alignment of icons for folders, default favicon (New Tab page) and favicons should not differ whether there is the overflow indicator or not.**

7. Preferences > Bookmarks Bar > choose `Text only`
    - **Bookmarks bar should decrease in height by a couple of pixels**



## Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests


- [ ] Adequate test coverage exists to prevent regressions
- [ ] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [ ] New files have MPL2 license header

petemill commented:
I've temporarily cherry-picked this to 0.22.x-single-webview until it is approved. If a change is required, that's ok since 0.22.x-single-webview is reset for each build until single-webview is merged to 0.22.x-release3

cezaraugusto commented on app/renderer/components/bookmarks/bookmarksToolbar.js:
```diff
@@ -196,23 +195,23 @@ class BookmarksToolbar extends React.Component {
     const currentWindow = state.get('currentWindow')
     const activeFrame = frameStateUtil.getActiveFrame(currentWindow) || Immutable.Map()
+    const bookmarkDisplayMode = bookmarkUtil.showOnlyFavicon()
+        ? 1
```
it's unclear to me what those numbers mean. IMO better to have them under https://github.com/brave/browser-laptop/blob/master/app/common/constants/settingsEnums.js and name them something like `bookmarkDisplayOptions`

petemill replied:
I don't think that would be appropriate. Those numbers are here purely so that this component (only) will re-render when the mode changes. Putting that in a separate file only spreads logic that is otherwise fully self-contained in to another module. Plus, this is not a settings enum. This value is simply a tri-state value for a prop so that we can compare if the next update has a different display mode. It's irrelevant what the value is, apart from textOnly, which is why that is specified as a prop with a more understandable name.

bsclifton replied:
if only a tri-state value, why not use a string value instead? 😄

petemill replied:
Scratch that... it was appropriate because there is already an enum that I didn't know about, and it's being called many many times for each action / render cycle, so I started using it and then went down a rabbit hole of optimizing it for readability and less state lookups... 🤷‍♀️  ✅

cezaraugusto commented on app/renderer/components/bookmarks/bookmarksToolbar.js:
```diff
@@ -196,23 +195,23 @@ class BookmarksToolbar extends React.Component {
     props.shouldAllowWindowDrag = !isWindows && windowState.shouldAllowWindowDrag(state, currentWindow, activeFrame, isFocused(state))
     props.toolbarBookmarks = bookmarksState.getBookmarksWithFolders(state, 0).take(110).map(item => item.get('key'))
-
+    props.textOnly = (bookmarkDisplayMode === 3)
```
if we follow the above comment here we could have

```
props.textOnly = (bookmarkDisplayMode === bookmarkDisplayOptions.textOnly)
```

or any other more meaningful name

petemill replied:
The purpose of this prop is already to make a readable name: `textOnly` as that value *is* used in other places in the component, whereas the exact value of `bookmarkDisplayMode` is not considered anywhere else.

cezaraugusto commented on app/renderer/components/bookmarks/bookmarksToolbar.js:
```diff
@@ -303,29 +316,29 @@ class BookmarksToolbar extends React.Component {
-          size={globalStyles.spacing.bookmarksToolbarOverflowButtonWidth}
-          iconClass={globalStyles.appIcons.angleDoubleRight}
-          onClick={this.onMoreBookmarksMenu}
+        <svg className={css(styles.bookmarksToolbar__overflowIndicator__icon)} xmlns='http://www.w3.org/2000/svg'viewBox='0 0 320 512'>
```
is there any benefit of inlining SVGs instead of requiring the file here? if perf is the same I'd prefer the latter for readability.

petemill replied:
Benefit is:
- no extra network request
- ability to style through css

Sure, it can be imported as a separate component, e.g.  `<BookmarksToolbarOverflowIndicatorIcon />` and since this icon is changing in the icon project and I'm using this technique for most new icons, then I'll be sure to favor not having the `<svg>` paths make the other components look messy. I'd prefer to make that change for that release if that's alright with everyone else.

bsclifton replied:
being able to style through CSS is a huge win- I've fought angrily with SVGs when used as a value for `background-image` (and similar use-cases)

cezaraugusto replied:
ya sorry missed the className prop here

cezaraugusto commented on app/renderer/components/bookmarks/bookmarksToolbar.js:
```diff
@@ -283,11 +284,23 @@ class BookmarksToolbar extends React.Component {
+        // which gets overriden by this `className` attribute here.
+        (this.hasHiddenKeys
+          ? ' ' + css(styles.bookmarksToolbar_hasOverflow)
+          : '')
```
**nit** can't we put it inside `style` instead?

```
style={{ 
  '--bookmarks-toolbar-overflow-indicator-visibility': this.hasHiddenKeys && 'visible',
  '--bookmarks-toolbar-overflow-indicator-width': this.hasHiddenKeys && `${globalStyles.spacing.bookmarksToolbarOverflowButtonWidth} !important`
}}
```

it wouldn't parse if result is `false` and IMO it's easier to read. Again this is **nit**

petemill commented:
![image](https://user-images.githubusercontent.com/741836/39550343-c1334f82-4e15-11e8-93d6-8e76d6621c29.png)

petemill commented:
Updated with changes and updated description to reflect that I ended up using and modifying the bookmarkUtil settings enum property. At this stage in the project I wouldn't normally consider changing that, but it sets that property up nicely for state-selection-memoization coming in #12105. Bookmark toolbar is a great candidate for that as if you have a full bookmarks bar, that's a lot of components and state selection repeating for every single action dispatch cycle (often many per second, when the data doesn't change anyway).

cezaraugusto commented:
master (0.24.x) fdb0943
0.23.x 65a83fd37
0.22.x-release3 66d6cef96
